// mdl-ext/es/utils/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var VK_TAB = 9;
var VK_ENTER = 13;
var VK_ESC = 27;
var VK_SPACE = 32;
var VK_PAGE_UP = 33;
var VK_PAGE_DOWN = 34;
var VK_END = 35;
var VK_HOME = 36;
var VK_ARROW_LEFT = 37;
var VK_ARROW_UP = 38;
var VK_ARROW_RIGHT = 39;
var VK_ARROW_DOWN = 40;

var ARIA_EXPANDED = 'aria-expanded';
var ARIA_HIDDEN = 'aria-hidden';
var ARIA_MULTISELECTABLE = 'aria-multiselectable';
var ARIA_SELECTED = 'aria-selected';

var IS_DIRTY = 'is-dirty';
var IS_DISABLED = 'is-disabled';
var IS_EXPANDED = 'is-expanded';
var IS_FOCUSED = 'is-focused';
var IS_INVALID = 'is-invalid';
var IS_UPGRADED = 'is-upgraded';
var DATA_UPGRADED = 'data-upgraded';

var MDL_RIPPLE = 'mdl-ripple';
var MDL_RIPPLE_COMPONENT = 'MaterialRipple';
var MDL_RIPPLE_EFFECT = 'mdl-js-ripple-effect';
var MDL_RIPPLE_EFFECT_IGNORE_EVENTS = 'mdl-js-ripple-effect--ignore-events';

exports.VK_TAB = VK_TAB;
exports.VK_ENTER = VK_ENTER;
exports.VK_ESC = VK_ESC;
exports.VK_SPACE = VK_SPACE;
exports.VK_PAGE_UP = VK_PAGE_UP;
exports.VK_PAGE_DOWN = VK_PAGE_DOWN;
exports.VK_END = VK_END;
exports.VK_HOME = VK_HOME;
exports.VK_ARROW_LEFT = VK_ARROW_LEFT;
exports.VK_ARROW_UP = VK_ARROW_UP;
exports.VK_ARROW_RIGHT = VK_ARROW_RIGHT;
exports.VK_ARROW_DOWN = VK_ARROW_DOWN;
exports.ARIA_EXPANDED = ARIA_EXPANDED;
exports.ARIA_HIDDEN = ARIA_HIDDEN;
exports.ARIA_MULTISELECTABLE = ARIA_MULTISELECTABLE;
exports.ARIA_SELECTED = ARIA_SELECTED;
exports.IS_DIRTY = IS_DIRTY;
exports.IS_DISABLED = IS_DISABLED;
exports.IS_EXPANDED = IS_EXPANDED;
exports.IS_FOCUSED = IS_FOCUSED;
exports.IS_INVALID = IS_INVALID;
exports.IS_UPGRADED = IS_UPGRADED;
exports.DATA_UPGRADED = DATA_UPGRADED;
exports.MDL_RIPPLE = MDL_RIPPLE;
exports.MDL_RIPPLE_COMPONENT = MDL_RIPPLE_COMPONENT;
exports.MDL_RIPPLE_EFFECT = MDL_RIPPLE_EFFECT;
exports.MDL_RIPPLE_EFFECT_IGNORE_EVENTS = MDL_RIPPLE_EFFECT_IGNORE_EVENTS;
